src/query_plan/operation.rs

```diff
@@ -1667,15 +2224,93 @@ impl NormalizedInlineFragmentSelection {
                 directives: Arc::new(inline_fragment.directives.clone()),
                 selection_id: SelectionId::new(),
             }),
-            selection_set: NormalizedSelectionSet::normalize_and_expand_fragments(
+            selection_set: NormalizedSelectionSet::from_selection_set(
                 &inline_fragment.selection_set,
                 fragments,
                 schema,
-                normalize_fragment_spread_option,
             )?,
         })
     }
 
+    pub(crate) fn from_fragment_spread_selection(
+        fragment_spread_selection: &Arc<NormalizedFragmentSpreadSelection>,
+    ) -> Result<NormalizedInlineFragmentSelection, FederationError> {
+        let fragment_spread_data = fragment_spread_selection.spread.data();
+        Ok(NormalizedInlineFragmentSelection {
+            inline_fragment: NormalizedInlineFragment::new(NormalizedInlineFragmentData {
+                schema: fragment_spread_data.schema.clone(),
+                parent_type_position: fragment_spread_data.type_condition_position.clone(),
+                type_condition_position: Some(fragment_spread_data.type_condition_position.clone()),
+                directives: fragment_spread_data.directives.clone(),
+                selection_id: SelectionId::new(),
+            }),
+            selection_set: fragment_spread_selection
+                .selection_set
+                .expand_all_fragments()?,
+        })
+    }
+
+    pub(crate) fn normalize(
+        &self,
+        _parent_type: &CompositeTypeDefinitionPosition,
+    ) -> NormalizedInlineFragmentSelection {
+        // TODO - will be fixed in subsequent PR
+        self.clone()
+    }
+
+    pub(crate) fn rebase_on(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        named_fragments: &NamedFragments,
+        schema: &ValidFederationSchema,
+        error_handling: &RebaseErrorHandlingOption,
+    ) -> Result<Option<NormalizedSelection>, FederationError> {
+        if &self.inline_fragment.data().schema == schema
+            && self.inline_fragment.data().parent_type_position == *parent_type
+        {
+            return Ok(Some(NormalizedSelection::InlineFragment(Arc::new(
+                self.clone(),
+            ))));
+        }
+        return if let Some(rebased_fragment) =
+            self.inline_fragment
+                .rebase_on(parent_type, schema, error_handling)?
+        {
+            let rebased_casted_type = rebased_fragment
+                .data()
+                .type_condition_position
+                .clone()
+                .unwrap_or(rebased_fragment.data().parent_type_position.clone());
+            if &rebased_casted_type == parent_type {
+                Ok(Some(NormalizedSelection::InlineFragment(Arc::new(
+                    NormalizedInlineFragmentSelection {
+                        inline_fragment: rebased_fragment,
+                        selection_set: self.selection_set.clone(),
```

Comment:
Here but also in some other places we clone a selection set onto the rebased fragment. Could it potentially point to the wrong schema?

Say we rebase an inline fragment from schema A to schema B. The inline fragment applies to a type `X` that exists in both schemas. The check at the start of `rebase_on` is false because the schemas are different, so we do the rebase. The fragment is rebased. Then on line 2284 we compare the rebased type name to the existing parent type name. They're both `X`, so we enter this branch, and we return `rebased_fragment` with the old, unrebased selection set.

I think this behaviour is different from JS because JS compares references to type definitions rather than type names. I think the correct behaviour for us here would be to check that both the type and the schema are the same in the rebased type?

Reply:
Indeed I missed the extra check whether we are rebasing on the same schema (i.e. it should be `same_schema && rebased_casted_type == parent_type`)

In all other places where are cloning the selection sets it was within the same schema.